fix(compiler): add validations for custom placeholder expressions

crisbeto opened the pull request:
Adds a few basic validations to custom placeholder expressions (e.g. `{{ foo // i18n(ph = "foo") }}`) in order to catch cases like newlines inside the placeholder and non-matching quotes.

Note that I've intentionally kept the validations somewhat basic for now, because the previous approach only used regex for the parsing which would've allowed other potentially invalid expressions. We can expand this approach if we run into more such cases in the future.

Fixes #45653.

crisbeto commented on packages/compiler/src/i18n/i18n_parser.ts:
```diff
@@ -276,9 +278,47 @@ function assertEquivalentNodes(previousNodes: i18n.Node[], nodes: i18n.Node[]):
+        continue;
+      }
+
+      throw new Error('Custom placeholder must be wrapped in quotes.');
```
This validation broke some targets, because some apps weren't wrapping the placeholders in quotes. Is this something we want to enforce? Technically it's outside the scope of the initial bug report, but I added it because it seemed reasonable. I'm not super attached to it though.

AndrewKushnir replied:
Internal (g3) i18n linter seems to allow placeholders without quotes, so we can drop this enforcement from the PR.

Having quotes around placeholder names makes sense to me (esp. in a context of adding more keys to the `i18n(...)` block in the future), so we can enforce that in a separate PR.

crisbeto replied:
I think that I'll keep it for now, pending the TGP. It looks like omitting the quotes actually prevented the placeholder from being matched so it would fall back to the default `INTERPOLATION` placeholder instead of the specified one.

crisbeto committed:
fix(compiler): add validations for custom placeholder expressions

Adds a few basic validations to custom placeholder expressions (e.g. `{{ foo // i18n(ph = "foo") }}`) in order to catch cases like newlines inside the placeholder and non-matching quotes.

Note that I've intentionally kept the validations somewhat basic for now, because the previous approach only used regex for the parsing which would've allowed other potentially invalid expressions. We can expand this approach if we run into more such cases in the future.

Fixes #45653.

dgp1130 commented on packages/compiler/src/i18n/i18n_parser.ts:
```diff
@@ -276,9 +278,47 @@ function assertEquivalentNodes(previousNodes: i18n.Node[], nodes: i18n.Node[]):
+    if (i === rawPlaceholder.length - 1) {
+      if (char === rawPlaceholder[0]) {
+        continue;
+      }
```
Nit: I think it would be cleaner to check the first and last characters are quotes prior to the for-loop and then slice the string and iterate like:

```typescript
const firstChar = rawPlaceholder[0];
if (firstChar !== '"' && firstChar !== '\'') throw new Error('...');
const lastChar = rawPlaceholder[rawPlaceholder - 1];
if (secondChar !== '"' && secondChar !== '\'') throw new Error('...');

const content = rawPlaceholder.slice(1 -1);
for (const char of content) {
  if (char === '\n' || char === '\r') throw new Error('...');
}

return content;
```

That removes the need to `continue` for the first and last character or track character indexes which are fairly unrelated to this logic. You also don't need a separate `result` value since it isn't being transformed, just validated.

crisbeto replied:
It might just be me, but I find that harder to follow.

dgp1130 replied:
In my mind, a for-loop is trying to repeat logic N times, which is completely counter to validating quotes because a), it happens exactly twice every time, and b) the action take is completely unrelated to the rest of the loop. The way it's currently written effectively says "Do this N times, except for the first character in which we do something completely different, and except for the last character in which we do something completely different again." When it's doing something completely different, a for-loop isn't really the right construct. There are of course exceptions like "every time you see a quote do something different", but it feels avoidable and unnecessary here which is why I called it out.

Personally I also find that the less indexes to manage and the less `continue` / `break` statements to worry about makes me less likely to screw something up and / or forget something, though that is probably a more subjective point.

Also manually appending to `result` gives an opportunity to mess it up (if you added an if statement in the future, it would be easy to accidentally append the character in only one branch), whereas treating the string as readonly after stripping quotes and only validating that it is correct means we can't introduce a bug where we validate one string correctly but then build an incorrect string as the result.

Everyone thinks differently ofc, and I don't mean to be too nitpicky, just something which stood out to me as having a clear alternative.

dgp1130 commented on packages/compiler/src/i18n/i18n_parser.ts:
```diff
@@ -276,9 +278,47 @@ function assertEquivalentNodes(previousNodes: i18n.Node[], nodes: i18n.Node[]):
+      throw new Error(`Custom placeholder must be a valid string in expression "{{${input}}}".`);
+    }
+
+    if (char === '\n' || char === '\r') {
```
Nit: Might be out of scope, but I'm wondering if we should allowlist the characters we want rather than blocklist the ones we don't. I imagine we don't want folks putting Klingon or emojis in here. Other forms of whitespace might be an issue too like tabs, no-break spaces, or zero-width spaces. Would it be worth limiting this to something like latin alphabet, digits, `-`, `_`, `$`, and space? That would probably cover 99% of use cases.

crisbeto replied:
I think that'll definitely break even more targets. I've seen placeholders where the author was writing things like "Your name (>= 2 characters)". IMO we shouldn't be too strict as to what can be inside the string since we don't really care about it downstream.

dgp1130 replied:
This does get turned into placeholders which are visible to translators and do have some requirements (for instance `i18n(ph = "user name")` becomes `USER_NAME`. I have no idea what "(>= 2 characters)" would become, and that kind of information should be in the description IMHO.

I'm not sure how strict JSCompiler is and we should probably align with that where we can.

crisbeto replied:
I think that it turns the characters it doesn't know about into underscores, but I'd have to double check. I agree that people seem to be misusing it, but I'm nervous about breaking even more translation strings.

dgp1130 replied:
Personally I think that's all the more reason we should be stricter about this since `YOUR_NAME____2_CHARACTERS` isn't communicating what the developer wanted it to. I agree the migration problem might be bigger than we want to chew on right now, and I'm happy to scope this down to make it more manageable if necessary. I'm just thinking it would be nice to have a fully validated end state we're working towards.

crisbeto replied:
Agreed. I'll keep the validation as it is for this PR and I'll do a second PR where I make the validation even stricter.

dgp1130 commented on packages/compiler/src/i18n/i18n_parser.ts:
```diff
@@ -276,9 +278,47 @@ function assertEquivalentNodes(previousNodes: i18n.Node[], nodes: i18n.Node[]):
 
-const _CUSTOM_PH_EXP =
-    /\/\/[\s\S]*i18n[\s\S]*\([\s\S]*ph[\s\S]*=[\s\S]*("|')([\s\S]*?)\1[\s\S]*\)/g;
+function extractPlaceholderName(input: string): string|null {
```
Might also be kind of out of scope for this PR, but doesn't the compiler have more comprehensive diagnostic utilities? Shouldn't this take in a node with a source span and return a diagnostic which links to that span so users get good syntax highlighting of the error location?

And won't throwing here crash the compiler, or at least abort detecting other errors in the same file? Would that cause any issue with the language service?

crisbeto replied:
I was wondering the same when I was writing it, but I saw that there are other places where the parser straight-up throws errors so I figured it's the way to go.

dgp1130 replied:
I'll defer to others more knowledgeable in how the compiler does this kind of thing.
